File: plugins/adaptive-learner-plugin-learning-repo/adaptive_learner_learning_repo/meta/roadmap.py
"""ROADMAP.md generator.

Surfaces the immediate next-step recommendation (resume active
session / start first session / start next) plus the list of
topics already in flight. Curriculum-level future topics
(``LearningTopic`` rows not yet touched by any session) are NOT
included here in commit 3 because the data model doesn't link
``LearningProject`` to ``Curriculum`` directly — that surface
arrives if + when the cross-link lands.
"""

from __future__ import annotations

from ..context import RenderContext
from ..labels import Labels


def render_roadmap(ctx: RenderContext, labels: Labels) -> str:
    """Render ROADMAP.md."""

    lines: list[str] = [
        f"# {labels.roadmap_title}",
        "",
        labels.roadmap_intro,
        "",
    ]
    lines.extend(_next_steps_block(ctx, labels))
    lines.extend(_open_topics_block(ctx, labels))
    return "\n".join(lines).rstrip() + "\n"


def _next_steps_block(ctx: RenderContext, labels: Labels) -> list[str]:
    lines = [f"## {labels.roadmap_next_steps_heading}", ""]
    suggestion = _suggest_next_step(ctx, labels)
    if suggestion is None:
        lines.extend([labels.roadmap_no_next_steps, ""])
        return lines
    lines.append(f"- {suggestion}")
    lines.append("")
    return lines


def _suggest_next_step(ctx: RenderContext, labels: Labels) -> str | None:
    """Pick the most actionable next-step suggestion."""

    active = _active_session(ctx)
    if active is not None:
        return labels.roadmap_resume_active.format(
            method=active.method,
            step=active.cycle_step,
            cycle=active.cycle_count,
        )
    last_completed = _last_completed_session(ctx)
    if last_completed is not None:
        return labels.roadmap_start_next.format(method=last_completed.method)
    if ctx.sessions:
        # All sessions abandoned, no active, no completed — fall
        # through to the "start first" suggestion using the
        # first session's method (best signal we have).
        first_method = min(ctx.sessions, key=lambda s: s.started_at).method
        return labels.roadmap_start_first.format(method=first_method)
    return None


def _active_session(ctx: RenderContext):
    return next(
        (s for s in ctx.sessions if s.status == "active"),
        None,
    )


def _last_completed_session(ctx: RenderContext):
    completed = [s for s in ctx.sessions if s.status == "completed"]
    if not completed:
        return None
    return max(completed, key=lambda s: s.ended_at or s.started_at)


def _open_topics_block(ctx: RenderContext, labels: Labels) -> list[str]:
    lines = [f"## {labels.roadmap_open_topics_heading}", ""]
    if not ctx.topics:
        lines.extend([labels.roadmap_no_open_topics, ""])
        return lines
    for topic in ctx.topics:
        methods = ", ".join(topic.methods) if topic.methods else "-"
        lines.append(f"- **{topic.title}** ({len(topic.session_ids)} sessions; methods: {methods})")
    lines.append("")
    return lines

File: frontend/src/lib/learning-repo/renderers/roadmap.ts
/**
 * ROADMAP.md generator (Phase 49D / v1.32.0 /
 * PHASE-42-STORAGE-ABSTRACTION-01).
 *
 * Mirrors the Python ``roadmap.py`` at
 * ``plugins/.../learning-repo/adaptive_learner_learning_repo/
 * meta/roadmap.py``. Parity test (49F) pins byte-for-byte.
 *
 * Surfaces:
 *   - Immediate next-step suggestion (resume-active /
 *     start-next / start-first) — picked from the most
 *     actionable signal in the project's session history.
 *   - Open topics in flight (with session count + methods).
 *
 * Curriculum-level future topics are NOT surfaced here per
 * the Python source: the data model doesn't link
 * ``LearningProject`` to ``Curriculum`` directly.
 */

import {formatLabel, type Labels} from "../labels";
import type {RenderContext, SessionData} from "../render-context";

/**
 * Render the ROADMAP.md body: an immediate next-step
 * suggestion (resume-active / start-next / start-first) plus
 * the list of open topics still in flight.
 */
export function renderRoadmap(
    ctx: RenderContext,
    labels: Labels,
): string {
    const lines: string[] = [
        `# ${labels.roadmap_title}`,
        "",
        labels.roadmap_intro,
        "",
    ];
    lines.push(...nextStepsBlock(ctx, labels));
    lines.push(...openTopicsBlock(ctx, labels));
    return rstripWithNewline(lines.join("\n"));
}

function nextStepsBlock(
    ctx: RenderContext,
    labels: Labels,
): string[] {
    const lines: string[] = [
        `## ${labels.roadmap_next_steps_heading}`,
        "",
    ];
    const suggestion = suggestNextStep(ctx, labels);
    if (suggestion === null) {
        lines.push(labels.roadmap_no_next_steps, "");
        return lines;
    }
    lines.push(`- ${suggestion}`, "");
    return lines;
}

function suggestNextStep(
    ctx: RenderContext,
    labels: Labels,
): string | null {
    const active = activeSession(ctx);
    if (active !== null) {
        return formatLabel(labels.roadmap_resume_active, {
            method: active.method,
            step: active.cycle_step,
            cycle: active.cycle_count ?? 1,
        });
    }
    const lastCompleted = lastCompletedSession(ctx);
    if (lastCompleted !== null) {
        return formatLabel(labels.roadmap_start_next, {
            method: lastCompleted.method,
        });
    }
    if (ctx.sessions.length > 0) {
        // All sessions abandoned (no active, no completed) →
        // fall through to "start first" using the
        // chronologically first session's method as best
        // signal.
        const sorted = [...ctx.sessions].sort((a, b) =>
            a.started_at.localeCompare(b.started_at),
        );
        return formatLabel(labels.roadmap_start_first, {
            method: sorted[0].method,
        });
    }
    return null;
}

function activeSession(ctx: RenderContext): SessionData | null {
    for (const s of ctx.sessions) {
        if (s.status === "active") return s;
    }
    return null;
}

function lastCompletedSession(ctx: RenderContext): SessionData | null {
    const completed = ctx.sessions.filter((s) => s.status === "completed");
    if (completed.length === 0) return null;
    // Python's ``max(..., key=lambda s: s.ended_at or s.started_at)``
    // — fall back to started_at when ended_at is null.
    return completed.reduce((best, s) => {
        const bestKey = best.ended_at ?? best.started_at;
        const sKey = s.ended_at ?? s.started_at;
        return sKey.localeCompare(bestKey) > 0 ? s : best;
    });
}

function openTopicsBlock(
    ctx: RenderContext,
    labels: Labels,
): string[] {
    const lines: string[] = [
        `## ${labels.roadmap_open_topics_heading}`,
        "",
    ];
    if (ctx.topics.length === 0) {
        lines.push(labels.roadmap_no_open_topics, "");
        return lines;
    }
    for (const topic of ctx.topics) {
        const methods =
            topic.methods.length === 0 ? "-" : topic.methods.join(", ");
        lines.push(
            `- **${topic.title}** (${topic.session_ids.length} sessions; methods: ${methods})`,
        );
    }
    lines.push("");
    return lines;
}

function rstripWithNewline(s: string): string {
    return s.replace(/[\s\n]+$/, "") + "\n";
}
